Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
+end
+
+if length(xQ.Initial_PLD)>1 && length(xQ.Initial_PLD) ~= nEncodedVolumes
+	error('Length of Initial_PLD vector does not match the number of volumes.');
+end
+
+% Prepare the parameters also for interleaved reordering
+% We thus calculate the size of each Hadamard block as the number of Hadamard phases and TEs
+nVolumesPerRepetition = xQ.TimeEncodedMatrixSize * xQ.NumberEchoTimes;
+nRepetitions = nEncodedVolumes / nVolumesPerRepetition; % Calculating no. of acquisition repeats accounting for multipleTE
+
+if length(xQ.EchoTime)>1 && length(xQ.EchoTime) ~= nEncodedVolumes
+	error('Length of EchoTime vector does not match the number of volumes.');
+end
+
+% Currently, we only implement the option that TE times are together
+% In case a different ordering is used, an error is reported
+if isfield(xQ, 'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
+	% The first two EchoTimes are equal
+	if length(uniquetol(xQ.EchoTime(1:2), 0.001)) == 1
+		error('Processing of multi-TE sequence is only implemented for the case of TEs being together in ASL4D.')
+	end
+end
+
+%% 3. Obtain control4D images
+% Here we select all TEs for the control images
 
-nEncodedVolumes = size(imEncoded, 4);
+% For example for 64 volumes and 2 repetitions with 8 PLDs and 4 TEs, it takes volume 1,2,3,4 and 33,34,35,36 to get all TEs of the control
+indexControl = ((1:xQ.NumberEchoTimes)'*ones(1,nRepetitions) + ones(xQ.NumberEchoTimes,1)*(0:nRepetitions-1)*nVolumesPerRepetition);
+indexControl = indexControl(:)';
 
-nRepetitions = nEncodedVolumes / (xQ.TimeEncodedMatrixSize * xQ.NumberEchoTimes); % Calculating no. of acquisition repeats
+decodedControl4D = imEncoded(:,:,:, indexControl);
+xQ.EchoTime_Control4D = xQ.EchoTime(indexControl);
+xQ.InitialPLD_Control4D = xQ.Initial_PLD(indexControl);
+xQ.LabelingDuration_Control4D = xQ.LabelingDuration(indexControl);
 
-%% 2. Reorder multi-TE data
+%% 4. Reorder image matrix
 % At this point the data is organized like this (in terms of ASL4D.nii volumes):
 % PLD1/TE1,PLD1/TE2,PLD1/TE3,PLD1/TE4...PLD2/TE1,PLD2/TE2,PLD2/TE3... (TEs in first dimension, PLDs after)
 %
 % And for decoding we want
 % TE1/PLD1,TE1/PLD2,TE1/PLD3,TE1/PLD4...TE2/PLD1,TE2/PLD2,TE2/PLD3,TE2/PLD4 (PLDs in the first dimension, TEs after)
-if isfield(xQ,'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
+% Then below, we apply this "transformation"/"decoding" vector to the image matrix and to the parameters TE/PLD/LD
+if isfield(xQ, 'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
 	% This is the original order
-	vectorOldOrder = 1:nEncodedVolumes;
+	vectorReorderEncoded = 1:nEncodedVolumes;
 	
 	% Shape to a matrix with TEs first and all the rest later
-	vectorOldOrder = reshape(vectorOldOrder, xQ.NumberEchoTimes, nEncodedVolumes/xQ.NumberEchoTimes);
+	vectorReorderEncoded = reshape(vectorReorderEncoded, xQ.NumberEchoTimes, nEncodedVolumes/xQ.NumberEchoTimes);
 	
 	% Flip the two dimensions and make a row vector again
```

Comment:
flipping == mirroring, but here you mean transposing from columns to rows right?

Reply:
I meant that we turn the dimension TE-PLD over to PLD-TE. But I agree that to `switch` dimensions is better than flip. I will change that.

Reply:
Matlab calls that shifting dims. So I would call it shifting (like shifting gears, from 3rd gear/dim into 4rd gear/dim)

Reply:
Shifting moves 2->3, 3->4 etc. We switch/swap 2->3 and 3->2.

All clarified in c44ef7a2